### PR TITLE
Created Unit of Work

### PR DESCRIPTION
**Pull request by waffillka**

### CODE REVIEW

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/HistoryParams.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Bookcrossing.Contracts.Abstractions.RequestFeatures
+{
+    public class HistoryParams : RequestFeatures
+    {
+        public string MatchString { get; set; }
+        public DateTime From { get; set; } = default;
+        public DateTime To { get; set; } = default;
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/HistoryParams.cs]: delete default here. 
They will have default value without specification.

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿namespace Bookcrossing.Contracts.Abstractions.RequestFeatures
+{
+    public abstract class RequestFeatures
+    {
+        private const int _maxPageSize = 100;
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs]: in most cases we use TitleCase for consts
https://stackoverflow.com/questions/242534/c-sharp-naming-convention-for-constants

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        public int PageNumber { get; set; } = 1;
+        private int _pageSize = 10;
+        public string OrderyBy { get; set; } = "abc";
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs]: what is "abc"?

File: Bookcrossing.Data/Configuration/DependenciesConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        public static IServiceCollection RegisterRepositories(this IServiceCollection services)
+        {
+            services.AddScoped<IAuthorRepository, AuthorRepository>();
+            services.AddScoped<IPublisherRepository, PublisherRepository>();
+            services.AddScoped<IHistoryOfIssuingBooksRepository, HistoryOfIssuingBooksRepository>();
+
+            services.AddScoped<IBookRepository, BookRepository>();
+            services.AddScoped<IRepositoryManager, RepositoryManager>();
+
+            return services;
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Configuration/DependenciesConfiguration.cs]: check this library https://github.com/khellang/Scrutor
could be quite useful if you have a lot of services/repositories

File: Bookcrossing.Data/DbContext/BookcrossingDbContext.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            : base(options)
+        {
+        }
+
+        public DbSet<Book> Books { get; set; }
+        public DbSet<Author> Authors { get; set; }
+        public DbSet<Publisher> Publishers { get; set; }
+        public DbSet<User> Users { get; set; }
+        public DbSet<HistoryOfIssuingBooks> HistoryOfIssuingBooks { get; set; }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
```
Comment (KorsakVlad) [inline Bookcrossing.Data/DbContext/BookcrossingDbContext.cs:19]: add empty row

File: Bookcrossing.Data/Entities/Author.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+using System.ComponentModel.DataAnnotations;
+
+namespace Bookcrossing.Data.Entities
+{
+    public class Author : EntityBase
+    {
+        [Required]
+        public string Name { get; set; }
+        public DateTime Birthday { get; set; }
+        public virtual ICollection<Book>? Books { get; set; }
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Entities/Author.cs]: delete useless '?' (related to all places where you did for collections or classes)
init all IColletions in ctor like => Books = new List<Book>()

File: Bookcrossing.Data/Repositories/BookRepository.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+        }
+
+        public async Task<IQueryable<Book>> GetAsync(BookParams parameters, CancellationToken ct = default)
+        {
+            IQueryable<Book> entities = _dbContext.Books;
+
+            parameters ??= new BookParams();
+
+            if (string.IsNullOrEmpty(parameters.MatchString))
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/BookRepository.cs]: missed '!'

File: Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public HistoryOfIssuingBooksRepository(BookcrossingDbContext dbContext)
+           : base(dbContext)
+        {
+
+        }
+
+        public Task<IQueryable<HistoryOfIssuingBooks>> GetAsync(HistoryParams parametrs, CancellationToken ct = default)
+        {
+            throw new NotImplementedException();
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs]: ?

File: Bookcrossing.Data/Repositories/AuthorRepository.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        public AuthorRepository(BookcrossingDbContext dbContext)
+            : base(dbContext)
+        {
+
+        }
+
+        public async Task<IQueryable<Author>> GetAsync(AuthorPublisherParams parameters, CancellationToken ct = default)
+        {
+            var entities = _dbContext.Authors.Where(x => x.Name.Contains(parameters.MatchString));
+
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/AuthorRepository.cs]: entities = p.OrderBy == Order.Asc ? entites.OderBy : entities.OrderByDesc

File: Bookcrossing.Data/Repositories/BookRepository.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            {
+                entities = entities.Where(x => x.Name.Contains(parameters.MatchString));
+            }
+
+            if (parameters.OrderyBy == "abc")
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/BookRepository.cs]: ternarn

File: Bookcrossing.Data/Repositories/PublisherRepository.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        {
+
+        }
+
+        public async Task<IQueryable<Publisher>> GetAsync(AuthorPublisherParams parameters, CancellationToken ct = default)
+        {
+            var entities = _dbContext.Publishers.Where(x => x.Name.Contains(parameters.MatchString));
+
+            if (parameters.OrderyBy == "abc")
+            {
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/PublisherRepository.cs]: .

**Review by KorsakVlad — CHANGES_REQUESTED** [review]

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/AuthorPublisherParams.cs [hunk 12]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Bookcrossing.Contracts.Abstractions.RequestFeatures
+{
+    public class AuthorPublisherParams : RequestFeatures
+    {
+        public string SearchStringh { get; set; }
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/AuthorPublisherParams.cs]: typo

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/HistoryParams.cs [hunk 13]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Bookcrossing.Contracts.Abstractions.RequestFeatures
+{
+    public class HistoryParams : RequestFeatures
+    {
+        public string SearchStringh { get; set; }
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/HistoryParams.cs]: typo

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs [hunk 14]
```diff
@@ -0,0 +1,23 @@
+﻿using Bookcrossing.Contracts.Enumeration;
+
+namespace Bookcrossing.Contracts.Abstractions.RequestFeatures
+{
+    public abstract class RequestFeatures
+    {
+        private const int MaxPageSize = 100;
+        public int PageNumber { get; set; } = 1;
+        private int _pageSize = 10;
+        public Ordery OrderyBy { get; set; } = Ordery.Asc;
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/RequestFeatures.cs]: order

File: Bookcrossing.Contracts/Enumeration/Ordery.cs [hunk 15]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Bookcrossing.Contracts.Enumeration
+{
+    public enum Ordery
```
Comment (KorsakVlad) [inline Bookcrossing.Contracts/Enumeration/Ordery.cs]: order

File: Bookcrossing.Data/Entities/Book.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        public virtual ICollection<Author> Authors { get; set; }
+
+        [NotMapped]
+        public bool Free => RecipientId == null;
+
+        public Book()
+        {
+            Authors = new List<Author>();
+        }
+
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Entities/Book.cs]: -

File: Bookcrossing.Data/Entities/Author.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class Author : EntityBase
+    {
+        [Required]
+        public string Name { get; set; }
+        public DateTime Birthday { get; set; }
+        public virtual ICollection<Book> Books { get; set; }
+
+        public Author()
+        {
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Entities/Author.cs:10]: https://stackoverflow.com/questions/150479/order-of-items-in-classes-fields-properties-constructors-methods

File: Bookcrossing.Data/Repositories/BookRepository.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+            if (!string.IsNullOrWhiteSpace(parameters.SearchStringh))
+            {
+                entities = entities.Where(x => x.Name.Contains(parameters.SearchStringh)
+                                                || x.Description.Contains(parameters.SearchStringh)
+                                                || x.Authors.Any(c => c.Name.Contains(parameters.SearchStringh)) // что-то похожее на 48 строке 
+                                                || x.ISBIN.Contains(parameters.SearchStringh)
+                                                || x.Publisher.Name.Contains(parameters.SearchStringh));
+            }
+
+            if (parameters.IsFree == true)
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/BookRepository.cs]: delete

File: Bookcrossing.Data/Repositories/BookRepository.cs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public async Task<IQueryable<Book>> GetAsync(BookParams parameters, CancellationToken ct = default)
+        {
+            if (parameters == null)
+            {
+                throw new ArgumentNullException(nameof(parameters));
+            }
+
+            IQueryable<Book> entities = _dbContext.Books;
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/BookRepository.cs]: protected DbContext { get; }

File: Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        public async Task<IQueryable<HistoryOfIssuingBooks>> GetAsync(HistoryParams parameters, CancellationToken ct = default)
+        {
+            if (parameters == null)
+            {
+                throw new ArgumentNullException(nameof(parameters));
+            }
+
+            IQueryable<HistoryOfIssuingBooks> entities = _dbContext.HistoryOfIssuingBooks;
+
+            if (parameters.From != default)
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs:29]: refactorning

File: Bookcrossing.Contracts/Abstractions/RequestFeatures/AuthorPublisherParams.cs (hunk 12 above)
Comment (waffillka) [inline Bookcrossing.Contracts/Abstractions/RequestFeatures/AuthorPublisherParams.cs]: Rename property

File: Bookcrossing.Data/Repositories/BookRepository.cs [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+            if (!string.IsNullOrWhiteSpace(parameters.SearchStringh))
+            {
+                entities = entities.Where(x => x.Name.Contains(parameters.SearchStringh)
+                                                || x.Description.Contains(parameters.SearchStringh)
+                                                || x.Authors.Any(c => c.Name.Contains(parameters.SearchStringh)) // что-то похожее на 48 строке
```
Comment (waffillka) [inline Bookcrossing.Data/Repositories/BookRepository.cs]: delete one of

File: Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs (hunk 20 above)
Comment (waffillka) [inline Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs:29]: refactoring

File: Bookcrossing.Data/Repositories/RepositoryManager.cs [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+{
+    public class RepositoryManager : IRepositoryManager
+    {
+        private readonly BookcrossingDbContext _dbContext;
+        private IAuthorRepository _authorsRepository;
+        private IBookRepository _bookRepository;
+        private IHistoryOfIssuingBooksRepository _historyRepository;
+        private IPublisherRepository _publisherRepository;
+
+        public RepositoryManager(BookcrossingDbContext dbContext)
```
Comment (waffillka) [inline Bookcrossing.Data/Repositories/RepositoryManager.cs]: put fire and sword
火と剣を委託する

**Review by KorsakVlad — CHANGES_REQUESTED** [review]

File: Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        public async Task<IQueryable<HistoryOfIssuingBooks>> GetAsync(HistoryParams parameters, CancellationToken ct = default)
+        {
+            if (parameters == null)
+            {
+                throw new ArgumentNullException(nameof(parameters));
+            }
+
+            IQueryable<HistoryOfIssuingBooks> entities = DbContext.HistoryOfIssuingBooks;
+
+            if (parameters.From != default || parameters.To != default)
```
Comment (KorsakVlad) [inline Bookcrossing.Data/Repositories/HistoryOfIssuingBooksRepository.cs]: seperate

**Pull request merged**